### PR TITLE
Stomp

### PR DESCRIPTION
**Pull request by BilalRifas**
Fixes #<Issue Number>

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Required Balo version update
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: examples/stomp/publisher.bal [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+        login: "guest",
+        passcode: "guest",
+        vhost: "/",
+        acceptVersion: "1.1"
+    });
+
+public function main() {
+    // Message is published within the transaction block.
+    // Each message sent should be received from other end else rollback
+    // to retry the message along with transaction Id.
```
Comment (riyafa) [inline examples/stomp/publisher.bal]: Transaction block??

File: examples/stomp/publisher.bal [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+        // This sends the Ballerina message to the stomp broker.
+        string message1 = "Hello World From Ballerina - 1";
+        string message2 = "Hello World From Ballerina - 2";
+        string destination1 = "/queue/test";
+        string destination2 = "/queue/new";
+        var broadcast1 = stompSender->send(message1,destination1);
+        var broadcast2 = stompSender->send(message2,destination2);
+}
```
Comment (riyafa) [inline examples/stomp/publisher.bal]: We need to add a new line at the end of files in all places

File: examples/stomp/publisher.bal [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    // Each message sent should be received from other end else rollback
+    // to retry the message along with transaction Id.
+
+        // This sends the Ballerina message to the stomp broker.
+        string message1 = "Hello World From Ballerina - 1";
+        string message2 = "Hello World From Ballerina - 2";
+        string destination1 = "/queue/test";
+        string destination2 = "/queue/new";
+        var broadcast1 = stompSender->send(message1,destination1);
+        var broadcast2 = stompSender->send(message2,destination2);
```
Comment (riyafa) [inline examples/stomp/publisher.bal]: What does the send method return and what does broadcast mean?

File: examples/stomp/publisher.out [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+Successfully connected to stomp broker
+Connect to: 61613
+CONNECTED
+server:RabbitMQ/3.6.15
+session:session-ayVnDSHIzcO5_iTmH9tafA
+heart-beat:0,0
+version:1.1
+
+# transaction has started
+Begin has been made with transaction: b632dfdb-e182-44bd-a2f7-66652129183c
```
Comment (riyafa) [inline examples/stomp/publisher.out]: We might have to reword this. This is not grammatically very correct.

File: examples/stomp/publisher.out [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+2019-02-21 11:01:33,882 INFO  [ballerina/transactions] - Created transaction: 7feaf167-de8e-446a-a71f-5982f82af488
+Message: Hello World From Ballerina is sent successfully
```
Comment (riyafa) [inline examples/stomp/publisher.out:15]: We might have to reword this as well

File: examples/stomp/publisher.out [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+RECEIPT
+receipt-id:abee3b5e-a3b6-4ae9-aeb2-fa127c840aaf
+
+# disconnecting from the broker port
+Disconnected from stomp broker successfully
+Leave from: 61613
+2019-02-21 11:01:38,960 INFO  [ballerina/transactions] - Running 2-phase commit for transaction: 7feaf167-de8e-446a-a71f-5982f82af488:$anon$.$1
```
Comment (riyafa) [inline examples/stomp/publisher.out]: Why is transactions involved here?

File: examples/stomp/stomp_durable_topic_subscriber.bal [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+// This binds the created consumer to the listener service.
+service stompListeners on consumerEndpoint  {
+    // This resource is invoked when a message is received.
+    // Message object only gives us the string message.
+    resource function onMessage(stomp:Message message) {
+        io:println("-------------- Ballerina onMessage triggered -------------");
+        var messageContent = message.getContent();
+
+        io:println("Message : ", messageContent);
+        //var messageAck = message->ack();
```
Comment (riyafa) [inline examples/stomp/stomp_durable_topic_subscriber.bal]: Shall we remove commented code in all places?

File: examples/stomp/stomp_durable_topic_subscriber.bal [hunk 8]
```diff
@@ -0,0 +1,39 @@
+// This is the service implementation for STOMP protocol.
+import ballerina/stomp;
+import ballerina/log;
+import ballerina/io;
```
Comment (riyafa) [inline examples/stomp/stomp_durable_topic_subscriber.bal]: In services we use only log and in main we use only io in examples. Shall we adhere to that?

File: stdlib/stomp/pom-nats-compare.xml [hunk 9]
```diff
@@ -0,0 +1,2621 @@
...
+            <dependency>
+                <groupId>org.quartz-scheduler</groupId>
+                <artifactId>quartz-jobs</artifactId>
+                <version>${quartz.version}</version>
+            </dependency>
+            <dependency>
+                <groupId>com.fasterxml.jackson.core</groupId>
+                <artifactId>jackson-core</artifactId>
+                <version>${jackson-core.version}</version>
+            </de
```
Comment (riyafa) [inline stdlib/stomp/pom-nats-compare.xml]: What is this pom file?

File: stdlib/stomp/src/main/ballerina/stomp/listener_endpoint.bal [hunk 10]
```diff
@@ -0,0 +1,66 @@
+import ballerina/io;
```
Comment (riyafa) [inline stdlib/stomp/src/main/ballerina/stomp/listener_endpoint.bal]: Shall we add license header to all the files?

File: stdlib/stomp/src/main/ballerina/stomp/listener_endpoint.bal [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+import ballerina/log;
+import ballerina/runtime;
+
+public type Listener object {
+
+  *AbstractListener;
+
+  public ListenerConfig config = { };
+
+    // TODO return the error for Init as well, from new release
```
Comment (riyafa) [inline stdlib/stomp/src/main/ballerina/stomp/listener_endpoint.bal]: Shall we fix all todos?

File: stdlib/stomp/src/main/ballerina/stomp/sender.bal [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+# + passcode - STOMP passcode.
+# + vhost - default stomp vhost.
+# + acceptVersion - 1.1.
+# + socketClient - socketConnection.
+# + endOfFrame - null octet.
+
+public type Sender client object {
+
+    public string host = "";
+    public int port = 0;
```
Comment (riyafa) [inline stdlib/stomp/src/main/ballerina/stomp/sender.bal]: Shall we make host and port as mandatory parameters as we cannot make a connection without these? Also other fields that are needed

File: stdlib/stomp/src/main/ballerina/stomp/sender.bal [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+
+    public function __init(ConnectionConfiguration stompConfig) returns error? {
+        self.config = stompConfig;
+        self.host = stompConfig.host;
+        self.port = stompConfig.port;
+        self.login = stompConfig.login;
+        self.passcode = stompConfig.passcode;
+        self.vhost = stompConfig.vhost;
+        self.acceptVersion = stompConfig.acceptVersion;
+        self.socketClient = new({
```
Comment (riyafa) [inline stdlib/stomp/src/main/ballerina/stomp/sender.bal:36]: Are you using this socketclient anymore since you are using a java library now?

File: stdlib/stomp/src/main/ballerina/stomp/sender.bal [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+        "passcode:" + stompConfig.passcode + "\n" +
+        "host:" + stompConfig.vhost + "\n" +
+        "\n" + self.endOfFrame;
+
+    byte[] payloadByte = connect.toByteArray("UTF-8");
+    // Send desired content to the server using the write function.
+    var writeResult = socketClient->write(payloadByte);
+    if (writeResult is error) {
+        io:println("Unable to write the connect frame", writeResult);
+    }
```
Comment (riyafa) [inline stdlib/stomp/src/main/ballerina/stomp/sender.bal:81]: We shouldn't be handling errors here. Errors should be propagated to the user in all places. Let us not print in the connector files.

File: stdlib/stomp/src/main/ballerina/stomp/sender.bal [hunk 15]
```diff
@@ -0,0 +1,174 @@
...
+}
+
+public remote function Sender.send(string message, string destination) returns error?{
+    socket:Client socketClient = self.socketClient;
+
+    // Generating unique id for message receipt.
+    string messageId = system:uuid();
+
+    // SEND frame to send message.
+    string send = "SEND" + "\n" + "destination:" + destination + "\n" + "receipt:" + messageId + "\n" + "redelivered:" + "false" + "\n" + "content-type:"+"text/plain" + "\n" + "\n" + message + "\n" + self.endOfFrame;
```
Comment (riyafa) [inline stdlib/stomp/src/main/ballerina/stomp/sender.bal]: Shall we break the line at 120 characters in all places?

File: stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/Ack.java [hunk 16]
```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright (c) 2018 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (riyafa) [inline stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/Ack.java]: Let us change the year to 2019

File: stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/Ack.java [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+import org.ballerinalang.model.values.BValue;
+import org.ballerinalang.natives.annotations.BallerinaFunction;
+import org.ballerinalang.natives.annotations.Receiver;
+
+import static org.ballerinalang.stdlib.stomp.StompConstants.*;
+
+/**
+ * Initialize the Acknowledge.
+ *
+ * @since 0.990.2
```
Comment (riyafa) [inline stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/Ack.java]: Latest version would be 0.995

File: stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/Ack.java [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+ */
+@BallerinaFunction(
+        orgName = "ballerina",
+        packageName = "stomp",
```
Comment (riyafa) [inline stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/Ack.java]: Shall we have constants for frequently used strings?

File: stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/Ack.java [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+            try {
+                client.ack(String.valueOf(messageId));
+                System.out.println("Successfully acknowledged");
+            } catch (StompException e) {
+                context.setReturnValues(StompUtils.getError(context, "Error on acknowledging the message"));
+            //
+            } catch (NullPointerException n){
+                context.setReturnValues(StompUtils.getError(context, "Null value found for Ack"));
+            }
+            //
```
Comment (riyafa) [inline stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/Ack.java]: Dangling comments?

File: stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/Ack.java [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    @Override
+    public void execute(Context context) {
+        @SuppressWarnings(StompConstants.UNCHECKED)
+        BMap<String, BValue> message = (BMap<String, BValue>) context.getRefArgument(0);
+        DefaultStompClient client = (DefaultStompClient) message.getNativeData(StompConstants.CONFIG_FIELD_CLIENT_OBJ);
+        BValue messageId = message.get(StompConstants.MSG_ID);
+
+        if (this.ackMode.equals("auto")|this.ackMode.equals(null)) {
```
Comment (riyafa) [inline stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/Ack.java]: Let us use `||`  and also I don't think it is needed to use `this`. Also let us fix formatting issues

File: examples/stomp/publisher.bal (hunk 1 above)
Comment (BilalRifas) [reply examples/stomp/publisher.bal]: Transaction related comments will be removed

File: examples/stomp/publisher.out (hunk 6 above)
Comment (BilalRifas) [reply examples/stomp/publisher.out]: Transaction related logs will be removed

File: examples/stomp/stomp_durable_topic_subscriber.bal (hunk 7 above)
Comment (BilalRifas) [reply examples/stomp/stomp_durable_topic_subscriber.bal]: Sure will do

File: stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/Ack.java (hunk 16 above)
Comment (BilalRifas) [reply stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/Ack.java]: Corrected everywhere

File: stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/Ack.java (hunk 17 above)
Comment (BilalRifas) [reply stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/Ack.java]: Corrected everywhere

File: examples/stomp/publisher.bal (hunk 2 above)
Comment (BilalRifas) [reply examples/stomp/publisher.bal]: Corrected everywhere

File: examples/stomp/subscriber.bal [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+        log:printError("An error occured", err = er);
+    }
+}
+
+@stomp:ServiceConfig{
+    destination:"/queue/new",
+    ackMode:"AUTO"
+}
+
+service stompListener2 on consumerEndpoint  {
```
Comment (riyafa) [inline examples/stomp/subscriber.bal:43]: Can we have better names rather than naming as 1 and 2 in examples?

File: stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/DefaultStompClient.java [hunk 22]
```diff
@@ -0,0 +1,188 @@
...
+            if (callerType.equals("string")) {
+                    Executor.submit(messageResource, new ResponseCallback(body, messageId), new HashMap<>(), null, new BString(body));
+            } else if (callerType.equals("ballerina/stomp:Message")) {
+                    msgObj.addNativeData(STOMP_MSG, body);
+                    msgObj.put(MSG_CONTENT_NAME, new BString(body));
+                    msgObj.put(MSG_DESTINATION, new BString(destination));
+                    msgObj.put(MSG_ID, new BString(messageId));
+                    msgObj.addNativeData(CONFIG_FIELD_CLIENT_OBJ, this);
+                    Executor.submit(messageResource, new ResponseCallback(body, messageId), new HashMap<>(), null, msgObj);
+            }
```
Comment (riyafa) [inline stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/DefaultStompClient.java]: We need to discuss databinding. Shall we discuss in a future meeting?

File: stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/DefaultStompClient.java [hunk 23]
```diff
@@ -0,0 +1,188 @@
...
+        private String messageId;
+
+        ResponseCallback(String message, String messageId) {
+            this.message = message;
+            this.messageId = messageId;
+        }
+
+        @Override
+        public void notifySuccess() {
+            log.debug("Success notification");
```
Comment (riyafa) [inline stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/DefaultStompClient.java]:
```suggestion
            if (logger.isDebugEnabled()) {
                log.debug("Successful completion");
            }
```

File: stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/DefaultStompClient.java [hunk 24]
```diff
@@ -0,0 +1,188 @@
...
+                    msgObj.put(MSG_ID, new BString(messageId));
+                    msgObj.addNativeData(CONFIG_FIELD_CLIENT_OBJ, this);
+                    Executor.submit(messageResource, new ResponseCallback(body, messageId), new HashMap<>(), null, msgObj);
+            }
+
+    }
+
+    private class ResponseCallback implements CallableUnitCallback {
+        private String message;
+        private String messageId;
```
Comment (riyafa) [inline stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/DefaultStompClient.java]: Is these parameters used anywhere in the class?

File: stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/DefaultStompClient.java [hunk 25]
```diff
@@ -0,0 +1,188 @@
...
+
+        @Override
+        public void notifyFailure(BError error) {
+            ErrorHandlerUtils.printError("error: " + BLangVMErrors.getPrintableStackTrace(error));
+        }
+    }
+
+    @Override
+    public void onReceipt(String receiptId) {
+    }
```
Comment (riyafa) [inline stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/DefaultStompClient.java]: shall we add a comment for why this method is empty?

File: stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/DefaultStompClient.java [hunk 26]
```diff
@@ -0,0 +1,188 @@
...
+            Executor.submit(errorResource, new ResponseCallback(message, description),
+                        null, null, getErrorSignatureParameters(errorResource, message));
+        } catch (BallerinaConnectorException c) {
+                log.error("Error while executing onError resource", c);
+        }
+    }
+
+    @Override
+    public void onCriticalError(Exception e) {
+        System.out.println("Error: " + e);
```
Comment (riyafa) [inline stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/DefaultStompClient.java]: Let us never use System.out.println

File: stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/DefaultStompClient.java [hunk 27]
```diff
@@ -0,0 +1,188 @@
...
+            String resourceName = subService.getResources()[count].getName();
+            if (resourceName.equals("onError")) {
+                Resource onErrorResource = subService.getResources()[count];
+                if (onErrorResource != null) {
+                    errResource = onErrorResource;
+                }
+            }
+        }
+        return errResource;
+    }
```
Comment (riyafa) [inline stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/DefaultStompClient.java]: Instead of having two methods that have duplicate code, can we use one method and pass the resource name as a parameter

File: stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/StompException.java [hunk 28]
```diff
@@ -0,0 +1,40 @@
...
+    private String message;
+
+    public StompException(String message) {
+        super(message);
+        this.message = message;
+    }
+
+    public String getMessage() {
+        return message;
+    }
```
Comment (riyafa) [inline stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/StompException.java]: Is there a need to return this message?

File: stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/StompException.java [hunk 29]
```diff
@@ -0,0 +1,40 @@
...
+ */
+
+package org.ballerinalang.stdlib.stomp;
+
+/**
+ * Initialize the server stomp endpoint.
+ *
+ * @since 0.990.2
+ */
+public class StompException extends Exception {
```
Comment (riyafa) [inline stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/StompException.java]: Shall we extend `BallerinaConnectorExcetption` or `BallerinaException` to be more specific?

File: stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/endpoint/tcp/server/InitListener.java [hunk 30]
```diff
@@ -0,0 +1,114 @@
...
+    private static String hostName;
+    private static Integer stompPort;
+
+    @Override
+    public void execute(Context context) {
+        try {
+            BMap<String, BValue> connection = (BMap<String, BValue>) context.getRefArgument(0);
+
+            this.validateURI(context);
+            String connectionURI = "tcp://" + this.userLogin + ":" + this.userPasscode + "@" + this.hostName + ":" + this.stompPort;
```
Comment (riyafa) [inline stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/endpoint/tcp/server/InitListener.java]: Line longer than 120 characters

File: stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/endpoint/tcp/server/InitListener.java [hunk 31]
```diff
@@ -0,0 +1,114 @@
...
+            } else {
+                log.error("Login field is null");
+            }
+            //this.userLogin = String.valueOf(endpointConfig.get(CONFIG_FIELD_LOGIN));
+
+            BString passcode = (BString) endpointConfig.get(CONFIG_FIELD_PASSCODE);
+            if (passcode != null) {
+                this.userPasscode = String.valueOf(passcode);
+            } else {
+                log.error("Passcode field is null");
```
Comment (riyafa) [inline stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/endpoint/tcp/server/InitListener.java]: IMHO not having a passcode is not an error. So we should have a default value. I think it can never be null because already we have empty string in bal code.

File: stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/StompUtils.java [hunk 32]
```diff
@@ -0,0 +1,75 @@
...
+
+    private static final String STOMP_ERROR_CODE = "{ballerina/stomp}StompError";
+    private static final String STOMP_ERROR = "StompError";
+
+    /**
+     * Get error struct.
+     *
+     * @param context Represent ballerina context
+     * @param errMsg  Error message
+     * @return Error struct
```
Comment (riyafa) [inline stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/StompUtils.java]: Let us not use the word `struct` anywhere as Ballerina no longer has structs

File: stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/DefaultStompClient.java (hunk 27 above)
Comment (BilalRifas) [reply stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/DefaultStompClient.java]: fixed

File: examples/stomp/publisher.bal (hunk 3 above)
Comment (BilalRifas) [reply examples/stomp/publisher.bal]: Send method returns error or null and broadcast means publishing message to broker. For better understanding will be changing that to publish.

File: stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/DefaultStompClient.java (hunk 22 above)
Comment (BilalRifas) [reply stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/DefaultStompClient.java]: Sure will do

File: stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/StompException.java (hunk 28 above)
Comment (BilalRifas) [reply stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/StompException.java]: Yes, it should be.

File: stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/DefaultStompClient.java (hunk 24 above)
Comment (BilalRifas) [reply stdlib/stomp/src/main/java/org/ballerinalang/stdlib/stomp/DefaultStompClient.java]: Not used and removed
